internal/store/db/mongo/store_test.go

```diff
@@ -12,4 +14,207 @@
  * the License.
  *******************************************************************************/
 
+// This test will only be executed if the tag mongoRunning is added when running
+// the tests with a command like:
+// go test -tags mongoRunning
+
 package mongo
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/models"
+
+	"github.com/google/uuid"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var TestHost = "localhost"
+var TestPort = 27017
+var TestTimeout = 5000
+var TestDatabaseName = "test"
+var TestBatchSize = 1337
+
+var TestValidNoAuthConfig = db.Configuration{
+	Type:         db.MongoDB,
+	Host:         TestHost,
+	Port:         TestPort,
+	Timeout:      TestTimeout,
+	DatabaseName: TestDatabaseName,
+	BatchSize:    TestBatchSize,
+}
+
+var TestTimeoutConfig = db.Configuration{
+	Type:         db.MongoDB,
+	Host:         TestHost,
+	Port:         TestPort,
+	Timeout:      1,
+	DatabaseName: TestDatabaseName,
+	BatchSize:    TestBatchSize,
+}
+
+var TestID = primitive.NewObjectID().Hex()
+var TestAppServiceKey = uuid.New().String()
+var TestPayload = []byte("brandon was here")
+var TestPipelinePosition = 0
+var TestVersion = "1"
+
+var TestStoredObject = models.NewStoredObject(TestID, TestAppServiceKey, TestPayload, TestPipelinePosition, TestVersion)
+
+func TestNewClient(t *testing.T) {
+	tests := []struct {
+		name          string
+		config        db.Configuration
+		expectedError bool
+	}{
+		{"Success, no auth", TestValidNoAuthConfig, false},
+		{"Timed out", TestTimeoutConfig, true},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := NewClient(test.config)
+
+			if test.expectedError && err == nil {
+				t.Error("Expected an error")
+				return
+			}
+
+			if !test.expectedError && err != nil {
+				t.Errorf("Unexpectedly encountered error: %s", err.Error())
+				return
+			}
+		})
+	}
+}
+
+func TestClient_CRUD(t *testing.T) {
+	tests := []struct {
+		name             string
+		config           db.Configuration
+		expectedError    bool
+		expectedErrorVal error
+	}{
+		{"Success, no auth", TestValidNoAuthConfig, false, nil},
```

Comment:
More tests for error conditions? Seems like you need a way to specified expected error for each step in the CRUD test. Otherwise you can't test error from just `removeFromStore` near the end. Also need why to force error, like bad ID.

Maybe ad the concept of CRUD step and slice for expected error per step? and something similar for the inputs?

Reply:
> Maybe ad the concept of CRUD step and slice for expected error per step? and something similar for the inputs?

Can you expand upon this? Not sure what you mean.

Reply:
Your current test is only good for happy path and an expected error on the first `Store` step.

If you define the CRUD test into named steps you can have maps indexed by the step names to have the expected error info, alternate bad ID for input, etc. Steps would be something like `Store, Retrieve, Update, UpdateRetry, Remove` . In the test code use the step name to retrieve the expected error info appropriate for that section of the test, i.e. `Update` when verifying the error from calling `Update()` or the ID to use when calling `Update()` to test for error.